Autogen Tags enum, and allow specifying tags while defining an op

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77313

1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

Differential Revision: [D36912769](https://our.internmc.facebook.com/intern/diff/D36912769)

Internal Changes will be coming soon

anjali411 committed:
Autogen Tags enum, and allow specifying tags while defining an op

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"




[ghstack-poisoned]

anjali411 commented on torch/library.h:
```diff
@@ -592,12 +594,17 @@ class TORCH_API Library final {
+
   template <typename Schema>
-  Library& def(Schema&& raw_schema) & {
+  Library& def(Schema&& raw_schema, std::unordered_set<Tags> tags = {}) & {
```
couldn't make it optional because the compiler gets confused with the def below. So, maybe tags should be made non optional at other places too

anjali411 commented on torchgen/gen.py:
```diff
@@ -481,7 +482,10 @@ class RegisterSchema:
         if not self.selector.is_native_function_selected(f):
             return None
-        return f"m.def({cpp_string(str(f.func))});\n"
+        if len(f.tags) == 0:
```
snippet from `RegisterSchema.cpp`:
```
  m.def("rename_(Tensor(a!) self, Dimname[]? names) -> Tensor(a!)", std::unordered_set<Tags>({inplace_view}));
  m.def("rename(Tensor(a) self, Dimname[]? names) -> Tensor(a)");
  m.def("align_to(Tensor(a) self, Dimname[] names) -> Tensor(a)");
```

eellison replied:
This is a separate issue... but we probably dont want the tags to specify information that are already present in the schema, e.g. `inplace_view`

anjali411 replied:
cc. @bdhirsh

bdhirsh replied:
`inplace_view` is actually pretty useful because you can't tell just from the aliasing info `(a!)` if the op mutates the tensor's data vs. metadata.

We can technically figure it out by noticing that `transpose(Tensor(a), int dim0, int dim1) -> Tensor(a)` and `transpose_(Tensor(a!), int dim0, int dim1) -> Tensor(a!)` have similar schemas that can be paired up (so the mutable variant of a view op is an "inplace view"). But the tag is actually pretty useful in the codegen- we can use it for error checking, and once it's set then we don't need the grouping info anymore to know if we have an inplace view.

anjali411 commented on torchgen/gen.py:
```diff
@@ -1829,6 +1834,14 @@ def gen_aten_interned_strings() -> Dict[str, str]:
+            [f"{tag}" for tag in valid_tags]
+            ) + "};\n")
+        }
+    core_fm.write("enum_tags.h", gen_tags_enum)
```
```
#pragma once

// @generated by torchgen/gen.py from enum_tags.h

// Enum of valid tags obtained from the entries in tags.yaml
enum Tags
{
view_copy,
inplace_view
};
```

ezyang replied:
you need a namesapce

anjali411 replied:
adding it to `at` namespace

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"




[ghstack-poisoned]

anjali411 commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -210,7 +211,7 @@ class TORCH_API OperatorEntry final {
   OperatorName name_;
   c10::optional<AnnotatedSchema> schema_;
-
+  c10::optional<std::unordered_set<Tags>> tags_;
```
TODO: add a method to get tags

eellison commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -98,7 +99,7 @@ class TORCH_API OperatorEntry final {
   // versa that is an error.  (Refcounting for the registrations is
   // handled in the OperatorHandle in Dispatcher)
-  void registerSchema(FunctionSchema&&, std::string&& debug);
+  void registerSchema(FunctionSchema&&, std::string&& debug, const c10::optional<std::unordered_set<Tags>>& tags = c10::nullopt);
```
just fwiw, I imagine that for the number of tags we'll end up having for each operator a `std::vector` will be faster than `unordered_set`.. maybe we never expect to look up a tag on the hot path and that's fine

anjali411 replied:
yeah you are right, I have a local change which makes it a vector

eellison reviewed (commented):
Cool ! Maybe use https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L11436 in this pr ? Or is `native_functions.yaml` integration out of scope here

ezyang commented on torchgen/gen.py:
```diff
@@ -1829,6 +1838,17 @@ def gen_aten_interned_strings() -> Dict[str, str]:
+    def gen_tags_enum() -> Dict[str, str]:
+        return {
+            "enum_of_valid_tags": (
+                "enum Tags\n{\n"
```
Use an enum class; with enum they get dumped in global namespace

bdhirsh replied:
you might also want to dump the actual enum declaration bit directly inside of the template. That way someone staring at the template has a better idea of what the file is for without having to jump to codegen

bdhirsh commented on torchgen/gen.py:
```diff
@@ -481,7 +482,14 @@ class RegisterSchema:
             return None
-        return f"m.def({cpp_string(str(f.func))});\n"
+        if len(f.tags) == 0:
+            return f"m.def({cpp_string(str(f.func))});\n"
```
Also small nit, but you can probably just avoid adding a special case in the codegen for `len(f.tags) == 0`, since we'll just stamp out an empty set into `m.def()`.

bdhirsh commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -210,7 +211,7 @@ class TORCH_API OperatorEntry final {
   c10::optional<AnnotatedSchema> schema_;
-
+  c10::optional<std::unordered_set<Tags>> tags_;
   std::array<KernelFunction, c10::num_runtime_entries> dispatchTable_;
```
So today, `schema_` is a `c10::optional` but it's considered an error for `schema_` to be none at runtime (and we have a `DanglingImpl` test to catch that here: https://github.com/pytorch/pytorch/blob/5762c7b25b06620f95e251f2bf5d4a1564f99d03/aten/src/ATen/core/dispatch/Dispatcher.cpp#L324.

We might want to maintain the same invariant for `tags_`, so anyone using it at runtime doesn't have to worry about checking for whether the optional is invalid

ezyang replied:
Actually, this suggests we may want to refactor the "def" info into a single struct, which has a single optional surrounding it. Otherwise, it's possible to accidentally have schema_ nullopt but tags not (or vice versa), which is illegal

anjali411 replied:
tags is no longer optional so I guess we don't need to do anything here

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tags (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.

Follow ups:
1. Autogenerate python bindings for Tags Enum.
2. Expose tags in `torch.ops` API

[ghstack-poisoned]

albanD commented on tools/build_variables.bzl:
```diff
@@ -969,7 +969,9 @@ def glob_libtorch_python_sources(gencode_pattern = ":generate-code[{}]"):
         "torch/csrc/autograd/generated/python_linalg_functions.cpp",
+        "torch/csrc/autograd/generated/enum_tags.cpp",
         "torch/csrc/autograd/generated/python_return_types.cpp",
+        "torch/csrc/autograd/generated/enum_tags.cpp",
```
nit: once

albanD commented on tools/build_variables.bzl:
```diff
@@ -969,7 +969,9 @@ def glob_libtorch_python_sources(gencode_pattern = ":generate-code[{}]"):
         "torch/csrc/autograd/generated/python_nn_functions.cpp",
         "torch/csrc/autograd/generated/python_fft_functions.cpp",
         "torch/csrc/autograd/generated/python_linalg_functions.cpp",
+        "torch/csrc/autograd/generated/enum_tags.cpp",
```
python_enum_tags for better consistency

albanD commented on tools/autograd/templates/enum_tags.cpp:
```diff
@@ -0,0 +1,12 @@
+
+namespace py = pybind11;
+namespace torch {
+    namespace autograd {
```
Not sure we want them in the autograd namespace tbh?

albanD commented on tools/autograd/gen_python_functions.py:
```diff
@@ -320,6 +320,19 @@ def gen(
+    def gen_tags_enum() -> Dict[str, str]:
+        return {
+            "enum_of_valid_tags": (
+                'py::enum_<Tags>(m, "Tags")'
```
Why not have all the static part directly in the template? That will remove the need for a lot of `\n` and linebreaks here.

anjali411 commented on torch/csrc/autograd/enum_tags.h:
```diff
@@ -0,0 +1,8 @@
+
+namespace torch {
+    namespace autograd {
+    void initEnumTags(PyObject* module);
```
doesn't feel good putting in `torch::autograd`. @albanD where should this go?

ezyang replied:
The directory structure in torch/csrc is irredeemably hopeless. Just do something that locally makes sense.

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tags (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.

Follow ups:
1. Autogenerate python bindings for Tags Enum.
2. Expose tags in `torch.ops` API

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tags (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.

Follow ups:
1. Autogenerate python bindings for Tags Enum.
2. Expose tags in `torch.ops` API

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API

TODO:
1. add test to validate we are getting correct tags

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API

TODO:
1. add test to validate we are getting correct tags

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API

TODO:
1. add test to validate we are getting correct tags

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API

TODO:
1. add test to validate we are getting correct tags

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API

TODO:
1. add test to validate we are getting correct tags

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API

TODO:
1. add test to validate we are getting correct tags

[ghstack-poisoned]

anjali411 commented:
@ezyang @albanD added a new test in `test_ops.py` to verify correctness for `torch.Tags.inplace_view`. Eventually we should expand this test to add tests for every newly added tag.

mruberry commented on test/test_ops.py:
```diff
@@ -1340,10 +1341,62 @@ def is_bit_set(x):
             torch.is_complex,
         )
 
+class WrapperToTestTags(torch.Tensor):
```
Comment for this test class

anjali411 replied:
added a comment!

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -147,7 +147,7 @@ void Dispatcher::deregisterLibrary_(const std::string& ns) {
 }
 
-RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema, std::string debug) {
+RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema, std::string debug, const std::vector<at::Tag>& tags) {
```
Taking a const ref to vector is suboptimal here. As you need an owning reference accepting the vector by value will avoid an unnecessary copy when the argument is a temporary.

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -338,6 +339,19 @@ class TORCH_API OperatorHandle {
   }
 
+  const std::vector<at::Tag>& getTags() const {
+    return operatorDef_->op.getTags();
```
Consider returning an ArrayRef instead, it makes less commitments about the internal representation and is harder to accidentally copy

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -69,6 +70,7 @@ void OperatorEntry::registerSchema(FunctionSchema&& schema, std::string&& debug)
   // NB: don't register schema until after we've checked everything!
   dispatchKeyExtractor_.registerSchema(schema);
   schema_ = AnnotatedSchema(std::move(schema), std::move(debug));
+  tags_ = tags;
```
If you take by value you can then std::move here

ezyang commented on torch/csrc/jit/runtime/operator.h:
```diff
@@ -160,6 +160,16 @@ struct TORCH_API Operator {
   }
 
+  std::vector<at::Tag> getTags() const {
+    return op_.fold<std::vector<at::Tag>>(
```
You sure you want a copy here? ArrayRef would work well here

anjali411 replied:
Yeah because we return an empty vector for JitOnlyOperators. c10::ArrayRef wouldn't work because we return a newly created empty array/vector

ezyang replied:
You can return an empty array ref without actually having a newly created vector, since the pointer in that case can be null

ezyang commented on test/test_ops.py:
```diff
@@ -1340,10 +1341,65 @@ def is_bit_set(x):
+# A simple wrapper tensor subclass that runs runs
+# correctness checks to ensure that operators have expected
+# tags based on their input and ouput tensor properties
+class WrapperToTestTags(torch.Tensor):
```
A mode seems like a better way to implement this

anjali411 replied:
agreed! updated

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 commented on test/test_ops.py:
```diff
@@ -1342,6 +1344,39 @@ def is_bit_set(x):
+
+        # TODO: extend this test to test ops with multiple outputs and ops like native_batch_norm.out
+        # which mutate not necessarily the first input.
+        if isinstance(rs, torch.Tensor) and isinstance(args[0], torch.Tensor):
```
ahhh because it might or might not be a view cc. @albanD

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

albanD commented on test/test_ops.py:
```diff
@@ -1342,6 +1345,64 @@ def is_bit_set(x):
+        if isinstance(args[0], torch.Tensor):
+            old_size = args[0].size()
+            old_stride = args[0].stride()
+            with no_dispatch():
```
The two `with no_dispatch` can be removed I think

albanD reviewed (approved):
SGTM, only the nits from above.
Also Ed's comments on using ArrayRef look valid?

anjali411 commented on test/test_public_bindings.py:
```diff
@@ -241,6 +241,10 @@ def test_no_new_bindings(self):
 
             "wait",
+            "Tag",
+            "inplace_view",
```
only Tag, not its members

albanD replied:
So we are exposing stuff that we shouldn't? I think we should fix that!

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1346,7 +1346,7 @@ void initJITBindings(PyObject* module) {
                         {op}, symbol, args, kwargs, true);
                   });
-              return func;
+              return py::make_tuple(func, py::cast(op->getTags().vec()));
```
TODO: prevent the extra copy here in a future PR

anjali411 commented:
@pytorchbot merge

anjali411 commented:
@pytorchbot merge

malfet commented:
@pytorchbot revert -m "Broke OSS buck builds, see https://hud.pytorch.org/pytorch/pytorch/commit/9476a78f3754aa122323b431c59360b254559d16" -c nosignal

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36912769).

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

Differential Revision: [D36912769](https://our.internmc.facebook.com/intern/diff/D36912769)

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

Differential Revision: [D36912769](https://our.internmc.facebook.com/intern/diff/D36912769)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36912769).

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

Differential Revision: [D36912769](https://our.internmc.facebook.com/intern/diff/D36912769)

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

Differential Revision: [D36912769](https://our.internmc.facebook.com/intern/diff/D36912769)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36912769).

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

Differential Revision: [D36912769](https://our.internmc.facebook.com/intern/diff/D36912769)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36912769).

albanD commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -205,12 +206,16 @@ class TORCH_API OperatorEntry final {
   OperatorName name_;
   c10::optional<AnnotatedSchema> schema_;
-
+  #ifndef C10_MOBILE
```
nit: would be good to have a comment here that explains why we're doing this (binary size).

albanD reviewed (approved):
Change looks good. Just a nit about adding an extra comment.

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

Differential Revision: [D36912769](https://our.internmc.facebook.com/intern/diff/D36912769)

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

Differential Revision: [D36912769](https://our.internmc.facebook.com/intern/diff/D36912769)

[ghstack-poisoned]

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

Differential Revision: [D36912769](https://our.internmc.facebook.com/intern/diff/D36912769)

[ghstack-poisoned]

anjali411 commented:
@pytorchbot merge -g

anjali411 commented:
@pytorchmergebot merge

anjali411 commented:
@pytorchbot merge

anjali411 committed:
Update on "Autogen Tags enum, and allow specifying tags while defining an op"


1. Autogenerates Enum Tag (containing all valid tags from tags.yaml).
2. Adds an unordered set of tags to OperatorEntry class.
3. Updates Library def to optionally take tags.
4. Updates codegen for registering ops to pass tags to def.
5. Autogenerates python bindings for Tag Enum.
6. Expose tags in `torch.ops` API
7. New test to validate correctness of torch.Tags.inplace_view

TODO:
Add docs for Tags https://github.com/pytorch/pytorch/runs/6600629359?check_suite_focus=true

Differential Revision: [D36912769](https://our.internmc.facebook.com/intern/diff/D36912769)

Internal Changes will be coming soon

[ghstack-poisoned]
